csharp/src/Microsoft.ML.OnnxRuntime/SessionOptions.shared.cs

```diff
@@ -422,8 +425,70 @@ public void RegisterCustomOpLibrary(string libraryPath)
         /// <param name="libraryHandle">out parameter, library handle</param>
         public void RegisterCustomOpLibraryV2(string libraryPath, out IntPtr libraryHandle)
         {
+            // NOTE: This is confusing due to the history.
+            // SessionOptions.RegisterCustomOpLibrary initially called NativeMethods.OrtRegisterCustomOpsLibrary
+            // and leaked the handle.
+            // SessionOptions.RegisterCustomOpLibraryV2 was added to resolve that by returning the handle.
+            // Later, NativeMethods.OrtRegisterCustomOpsLibrary_V2 was added with ORT owning the handle.
+            //
+            // End result of that is
+            //   SessionOptions.RegisterCustomOpLibrary calls NativeMethods.OrtRegisterCustomOpsLibrary_V2
+            //   SessionOptions.RegisterCustomOpLibraryV2 calls NativeMethods.OrtRegisterCustomOpsLibrary
             var utf8Path = NativeOnnxValueHelper.StringToZeroTerminatedUtf8(libraryPath);
-            NativeApiStatus.VerifySuccess(NativeMethods.OrtRegisterCustomOpsLibrary(handle, utf8Path, out libraryHandle));
+            NativeApiStatus.VerifySuccess(NativeMethods.OrtRegisterCustomOpsLibrary(handle, utf8Path, 
+                                                                                    out libraryHandle));
+        }
+
+        /// <summary>
+        /// Register custom ops by calling the C function with functionName. 
+        /// The C function has the signature
+        ///   OrtStatus* RegisterCustomOps(OrtSessionOptions* options, const OrtApiBase* api);
+        /// 
+        /// The function must be available in the global symbols so ONNX Runtime can find it using GetProcAddress or
+        /// dlsym. This will require your app to 'DllImport' the native library containing the function and use 
+        /// Marshal.Prelink to ensure the function is loaded.
+        ///
+        /// Example usage:
+        ///   static class NativeCustomOps {
```

Comment:
It looks the benefit of creating a class NativeCustomOps, instead of RegisterCustomLibraryV2 is that user can use different RegisterCustomOpsFunctions function name, is there any use case for that?

Reply:
2 benefits. 

It provides a more generic way to setup the custom ops. Symbol just needs to be findable and the user can choose how they want that to happen. This can be used by any ORT user - it's not specific to extensions and how that is setup. Doesn't have to be a separate library. Most importantly, iOS doesn't have library paths at all so RegisterCustomLibraryV2 will never work on iOS.

It also lets different names be used, which we'd need if we split the extensions custom ops into different domains due to size and the user wanted to use multiple of those as we'd need a different registration function. e.g. on iOS both of those would be static libraries, and only one of the 2 RegisterCustomOps function definitions would be available (assuming it didn't fail at link time due to duplicate symbols).